### PR TITLE
istioctl: support `pc ecds` command

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

this PR will add new sub command `ecds` under `proxy-config`

dependent on upstream PR https://github.com/envoyproxy/envoy/pull/24384, **DO NOT BACKPORT**.

### CODE REVIEW

File: releasenotes/notes/42365.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: feature
+area: istioctl
+
+releaseNotes:
+- |
+  **Added** support for retrieves typed extension configuration for the Envoy in the specified pod by running `istioctl proxy-config ecds`.
```
Comment (dhawton) [inline releasenotes/notes/42365.yaml:7]: **Added** support for retrieving typed extension configuration from Envoy in the specified pod by running `istioctl proxy-config ecds`.

**Review by dhawton — CHANGES_REQUESTED** [review]

File: releasenotes/notes/42365.yaml (hunk 1 above)
Comment (zirain) [reply releasenotes/notes/42365.yaml]: updated

File: istioctl/pkg/writer/envoy/configdump/ecds.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		err := config.GetEcdsFilter().UnmarshalTo(c)
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve TypedExtensionConfig: %v", err)
+		}
+
+		ecds = append(ecds, c)
+	}
+
+	sort.Slice(ecds, func(i, j int) bool {
+		return ecds[i].GetName() < ecds[j].GetName()
```
Comment (howardjohn) [inline istioctl/pkg/writer/envoy/configdump/ecds.go:93]: should we also sort by type

File: istioctl/pkg/writer/envoy/configdump/xds.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+//go:generate sh -c "echo '// limitations under the License.\n' >> filter_types.gen.go"
+//go:generate sh -c "echo '//  GENERATED FILE -- DO NOT EDIT\n' >> filter_types.gen.go"
+//go:generate sh -c "echo 'package configdump\n\nimport (' >> filter_types.gen.go"
+//go:generate sh -c "go list github.com/envoyproxy/go-control-plane/... | grep 'v[2-9]' | grep -v /pkg/ | xargs -I{} echo '\t_ \"{}\"' >> filter_types.gen.go"
+//go:generate sh -c "echo '\n\t// Istio-specific Envoy filters' >> filter_types.gen.go"
+//go:generate sh -c "go list istio.io/api/envoy/config/filter/... | grep 'v[2-9]' | xargs -I{} echo '\t_ \"{}\"' >> filter_types.gen.go"
+//go:generate sh -c "echo ')' >> filter_types.gen.go"
+package configdump
+
+// Import all Envoy filter types so they are registered and deserialization does not fail
```
Comment (howardjohn) [inline istioctl/pkg/writer/envoy/configdump/xds.go:40]: why do we need this? can we just import `https://github.com/istio/istio/pull/42365#`?

File: istioctl/pkg/writer/envoy/configdump/testdata/ecds/output.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+NAME
```
Comment (howardjohn) [inline istioctl/pkg/writer/envoy/configdump/testdata/ecds/output.txt:1]: can we include the type? Possibly shortened for well known types, at least stripping `type.googleapis.com`.

We are going to use ECDS for more than wasm

File: releasenotes/notes/42365.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+apiVersion: release-notes/v2
+kind: feature
+area: istioctl
+
+releaseNotes:
+- |
+  **Added** support for retrieving typed extension configuration from Envoy in the specified pod by running `istioctl proxy-config ecds`.
```
Comment (ericvn) [inline releasenotes/notes/42365.yaml:7]: Maybe:
```suggestion
  **Added** `istioctl proxy-config ecds` to support retrieving typed extension configuration from Envoy for a specified pod.
```

File: istioctl/pkg/writer/envoy/configdump/ecds.go (hunk 2 above)
Comment (zirain) [reply istioctl/pkg/writer/envoy/configdump/ecds.go:93]: do you mean the real type of filter, for example, `type.googleapis.com/envoy.extensions.filters.http.wasm.v3.Wasm`?

File: istioctl/pkg/writer/envoy/configdump/xds.go (hunk 3 above)
Comment (zirain) [reply istioctl/pkg/writer/envoy/configdump/xds.go:40]: I import this for some proto unmarshal issue, let me find out if these still needed.

Comment (howardjohn) [reply istioctl/pkg/writer/envoy/configdump/xds.go:40]: Sorry I pasted wrong thing. I mean just import pkg/config/xds/filter_types.gen.go, don't duplicate it

Comment (zirain) [reply istioctl/pkg/writer/envoy/configdump/xds.go:40]: yes, that's the right way.

Comment (zirain) [reply istioctl/pkg/writer/envoy/configdump/xds.go:40]: we don't need this anymore for using `protojson`

File: istioctl/cmd/proxyconfig.go [hunk 6]
```diff
@@ -1280,3 +1281,57 @@ func getPodNameBySelector(labelSelector string) ([]string, string, error) {
 	ns = pl.Items[0].Namespace
 	return podNames, ns, nil
 }
+
+func ecdsConfigCmd() *cobra.Command {
+	var podName, podNamespace string
+
+	ecdsConfigCmd := &cobra.Command{
+		Use:   "ecds [<type>/]<name>[.<namespace>]",
```
Comment (hzxuzhonghu) [inline istioctl/cmd/proxyconfig.go:1289]: `ecds` or  `ec`

File: istioctl/cmd/proxyconfig.go [hunk 7]
```diff
@@ -1280,3 +1281,57 @@ func getPodNameBySelector(labelSelector string) ([]string, string, error) {
...
+  istioctl proxy-config ecds <pod-name[.namespace]>
+
+  # Retrieve endpoint summary without using Kubernetes API
+  ssh <user@hostname> 'curl localhost:15000/config_dump' > envoy-config.json
+  istioctl proxy-config ecds --file envoy-config.json
+`,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if (len(args) == 1) != (configDumpFile == "") {
+				cmd.Println(cmd.UsageString())
+				return fmt.Errorf("eds requires pod name or --file parameter")
```
Comment (hzxuzhonghu) [inline istioctl/cmd/proxyconfig.go:1302]: eds?

File: istioctl/cmd/proxyconfig.go (hunk 6 above)
Comment (zirain) [reply istioctl/cmd/proxyconfig.go]: update aliases

File: istioctl/pkg/writer/envoy/configdump/testdata/ecds/output.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+ECDS NAME                         Type
```
Comment (howardjohn) [inline istioctl/pkg/writer/envoy/configdump/testdata/ecds/output.txt:1]: Type -> TYPE would make other usages

**Review by howardjohn — APPROVED** [review]

**Review by hanxiaop — APPROVED** [review]

**Review by hzxuzhonghu — APPROVED** [review]

**Review by dhawton — APPROVED** [review]

**Pull request merged**